Mathlib/Algebra/Star/Positive.lean

```diff
@@ -0,0 +1,183 @@
+form `star s * s`. Even though this technically corresponds to the nonnegative elements, the
+"positive" terminology is standard throughout the literature (at least in C⋆-algebra theory and
+operator theory), so we choose to prefer that.
+
+The advantage of using this over other definitions of positivity is that it allows us to unify
+several different concepts under a single umbrella. For example, one might be tempted to consider
+the collection of bounded linear operators on an Hilbert space and define positivity as
+`∀ x, 0 ≤ ⟪T x, x⟫`, but this doesn't generalize nicely to C⋆-algebras. Alternatively, one could
+define positivity as those self-adjoint elements with nonnegative spectrum, but this requires the
+algebra to be unital. Our definition herein conveniently sidesteps these issues.
+
+We place everything in the `StarOrderedRing` namespace in order to avoid collisions or confusion
+with other parts of the library.
+
+## Main definitions
+
+* `StarOrderedRing.positive R`: The additive submonoid of a star ordered ring `R` generated by
+  elements of the form `star s * s` for `s : R`.
+* `StarOrderedRing.positive.selfAdjoint : positive R →+ selfAdjoint R`: the identity map as an
+  `AddMonoidHom`.
+* `StarOrderedRing.positive.instAddMonoidWithOne`: Makes `positive R` into an `AddMonoidWithOne`
+* `StarOrderedRing.positive.instCanonicallyOrderedAddMonoid`: Makes `positive R` into a
+  `CanonicallyOrderedAddMonoid` which avoids duplication of some lemmas (e.g., `zero_le`)
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+namespace StarOrderedRing
+
+/-- The positive elements of a star ordered ring, as an additive submonoid.
+
+Note that the type class assumptions for `StarOrderedRing.positive` are actually significantly
+weaker (just `OrderedAddCommMonoid`) so that it can be used in as many contexts as possible, but
+the primary intended use case is when `R` is a `StarOrderedRing`.  -/
+def positive (R : Type u) [OrderedAddCommMonoid R] : AddSubmonoid R where
+  carrier := {x : R | 0 ≤ x}
+  add_mem' := add_nonneg
+  zero_mem' := le_rfl
+
+theorem positive_def [OrderedAddCommMonoid R] {x : R} :
+    x ∈ positive R ↔ 0 ≤ x :=
+  Iff.rfl
+
+section NonUnitalSemiring
+
+variable [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+
+theorem positive_iff {x : R} :
+    x ∈ positive R ↔ x ∈ AddSubmonoid.closure (Set.range fun s : R => star s * s) := by
+  rw [positive_def, nonneg_iff]
+
+theorem positive_coe_set : (positive R : Set R) = {x | 0 ≤ x} := rfl
+
+theorem isSelfAdjoint_of_nonneg [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+    {x : R} (hx : 0 ≤ x) : IsSelfAdjoint x := by
+  rw [nonneg_iff] at hx
+  refine AddSubmonoid.closure_induction hx ?_ (isSelfAdjoint_zero R) fun _ _ => IsSelfAdjoint.add
+  rintro - ⟨s, rfl⟩
+  exact IsSelfAdjoint.star_mul_self s
+
+theorem le_iff_exists_positive {x y : R} : x ≤ y ↔ ∃ p : positive R, y = x + p := by
+  rw [le_iff']
+  constructor
```

Comment:
Is this `Subtype.exists` from here?

Reply:
Thanks! I was looking for that and didn't find it, but now I realized I could have just `simp`ed to get there.